src/main/java/io/confluent/connect/jdbc/dialect/DatabaseDialectProvider.java

```diff
@@ -0,0 +1,215 @@
+   * the subprotocol.
+   */
+  public interface JdbcUrlInfo {
+
+    /**
+     * Get the subprotocol in the JDBC URL.
+     *
+     * @return the subprotocol
+     */
+    String subprotocol();
+
+    /**
+     * Get the subname in the JDBC URL, which is everything after the ':' character following the
+     * subprotocol.
+     *
+     * @return the subname
+     */
+    String subname();
+
+    /**
+     * Get the full JDBC URL.
+     *
+     * @return the URL.
+     */
+    String url();
+  }
+
+  private final String name;
+
+  protected DatabaseDialectProvider(String name) {
+    assert name != null;
+    this.name = name;
+  }
+
+  /**
+   * The score that can be returned if the dialect does not match the connection.
+   */
+  public static final int NO_MATCH_SCORE = 0;
+
+  /**
+   * The minimum score that should be returned by the {@link #score(JdbcUrlInfo)} method to signal
+   * that the DatabaseDialect instance can be used with a connection.
+   */
+  public static final int MINIMUM_MATCHING_SCORE = 1;
+
+  /**
+   * The "average" score that should be returned by the {@link #score(JdbcUrlInfo)} method to signal
+   * that the DatabaseDialect instance can be used with a connection and .
+   */
+  public static final int AVERAGE_MATCHING_SCORE = 10;
+
+  /**
+   * The score for a dialect that is an excellent match.
+   */
+  public static final int PERFECT_MATCHING_SCORE = 1000;
+
+  /**
+   * Return the score describing how well this {@link DatabaseDialect} instance works with the given
+   * JDBC URL and {@link Connection}. The DatabaseDialect that scores the highest for a given
+   * connection will be used.
+   *
+   * @param jdbcInfo the information about the JDBC connection URL; may not be null
+   * @return the score; may be less than {@link #MINIMUM_MATCHING_SCORE} if the dialect does not
+   *     apply, or greater than or equal to {@link #MINIMUM_MATCHING_SCORE} if the DatabaseDialect
+   *     can be used with this connection.
+   */
+  public abstract int score(JdbcUrlInfo jdbcInfo);
+
+  /**
+   * Create a dialect instance.
+   *
+   * @param config the connector configuration
+   * @return the dialect instance; never null
+   */
+  public abstract DatabaseDialect create(AbstractConfig config);
+
+  /**
+   * Return the name of the dialect.
+   *
+   * @return the name of the dialect; may not be null
+   */
+  public String dialectName() {
+    return name;
+  }
+
+  @Override
+  public String toString() {
+    return dialectName();
+  }
+
+  /**
+   * An abstract base class for a provider that will return a
+   * {@link DatabaseDialectProvider#PERFECT_MATCHING_SCORE}
+   * if any of the supplied JDBC subprotocols match the subprotocol from the JDBC connection URL, or
+   * {@link DatabaseDialectProvider#NO_MATCH_SCORE} otherwise.
+   */
+  public abstract static class SubprotocolBasedProvider extends DatabaseDialectProvider {
```

Comment:
Here's an abstract subclass of `DatabaseDialectProvider` that we used in all of the DBMS-specific dialect providers. Simply provide the subprotocols, and URLs that use one of those subprotocols will be scored with a "perfect" score; anything else gets a "no match" score.